src/goto-programs/remove_const_function_pointers.cpp

```diff
@@ -0,0 +1,1013 @@
+  const exprt &expr, expressionst &out_resolved_expression, bool &out_is_const)
+{
+  const exprt &simplified_expr=simplify_expr(expr, ns);
+  bool resolved;
+  expressionst resolved_expressions;
+  bool is_resolved_expression_const;
+  if(simplified_expr.id()==ID_index)
+  {
+    const index_exprt &index_expr=to_index_expr(simplified_expr);
+    resolved=
+      try_resolve_index_of(
+        index_expr, resolved_expressions, is_resolved_expression_const);
+  }
+  else if(simplified_expr.id()==ID_member)
+  {
+    const member_exprt &member_expr=to_member_expr(simplified_expr);
+    resolved=try_resolve_member(
+      member_expr, resolved_expressions, is_resolved_expression_const);
+  }
+  else if(simplified_expr.id()==ID_dereference)
+  {
+    const dereference_exprt &deref=to_dereference_expr(simplified_expr);
+    resolved=
+      try_resolve_dereference(
+        deref, resolved_expressions, is_resolved_expression_const);
+  }
+  else if(simplified_expr.id()==ID_typecast)
+  {
+    typecast_exprt typecast_expr=to_typecast_expr(simplified_expr);
+    resolved=
+      try_resolve_typecast(
+        typecast_expr, resolved_expressions, is_resolved_expression_const);
+  }
+  else if(simplified_expr.id()==ID_symbol)
+  {
+    LOG("Non const symbol will not be squashed", simplified_expr);
+    resolved=false;
+  }
+  else
+  {
+    resolved_expressions.push_back(simplified_expr);
+    is_resolved_expression_const=is_const_expression(simplified_expr);
+    resolved=true;
+  }
+
+  if(resolved)
+  {
+    out_resolved_expression.insert(
+      out_resolved_expression.end(),
+      resolved_expressions.begin(),
+      resolved_expressions.end());
+    out_is_const=is_resolved_expression_const;
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::try_resolve_index_value
+
+  Inputs:
+   expr - The expression of the index of the index expression (e.g.
+          index_exprt::index())
+   out_array_index - The constant value the index takes
+
+ Outputs: Returns true if was able to find a constant value for the index
+          expression. If true, then out_array_index will be the index within
+          the array that the function pointer is pointing to.
+
+ Purpose: Given an index into an array, resolve, if possible, the index
+          that is being accessed. This deals with symbols and typecasts to
+          constant values.
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::try_resolve_index_value(
+  const exprt &expr, mp_integer &out_array_index)
+{
+  expressionst index_value_expressions;
+  bool is_const=false;
+  bool resolved=try_resolve_expression(expr, index_value_expressions, is_const);
+  if(resolved)
+  {
+    if(index_value_expressions.size()==1 &&
+      index_value_expressions.front().id()==ID_constant)
```

Comment:
You mean the line should be indented 3 more than the line above? I thought the guidelines said always two more than the previous line?